text/0000-projections-lifetimes-and-wf.md

```diff
@@ -0,0 +1,972 @@
+      R ⊢ 'x: 'a
+
+#### Outlives for type parameters
+
+For type parameters, the only way to draw "outlives" conclusions is to
+find information in the environment (which is being threaded
+implicitly here, since it is never modified). In terms of a Rust
+program, this means both explicit where-clauses and implied bounds
+derived from the signature (discussed below).
+
+    OutlivesTypeParameterEnv:
+      X: 'a in Env
+      --------------------------------------------------
+      R ⊢ X: 'a
+
+
+#### Outlives for projections
+
+Projections have the most possibilities. First, we may find
+information in the environment, as with type parameters, but we can
+also consult the trait definition to find bounds (consider an
+associated type declared like `type Foo: 'static`). These rule only
+apply if there are no higher-ranked lifetimes in the projection; for
+simplicity's sake, we encode that by requiring an empty list of
+higher-ranked lifetimes. (This is somewhat stricter than necessary,
+but reflects the behavior of my prototype implementation.)
+
+    OutlivesProjectionEnv:
+      <P0 as Trait<P1..Pn>>::Id: 'a in Env
+      --------------------------------------------------
+      <> ⊢ <P0 as Trait<P1..Pn>>::Id: 'a
+
+    OutlivesProjectionTraitDef:
+      WC = [Xi => Pi] WhereClauses(Trait) 
+      <P0 as Trait<P1..Pn>>::Id: 'a in WC
+      --------------------------------------------------
+      <> ⊢ <P0 as Trait<P1..Pn>>::Id: 'a
+
+All the rules covered so far already exist today. This last rule,
+however, is not only new, it is the crucial insight of this RFC.  It
+states that if all the components in a projection's trait reference
+outlive `'a`, then the projection must outlive `'a`:
+
+    OutlivesProjectionComponents:
+      ∀i. R ⊢ Pi: 'a
+      --------------------------------------------------
+      R ⊢ <P0 as Trait<P1..Pn>>::Id: 'a
+
+Given the importance of this rule, it's worth spending a bit of time
+discussing it in more detail. The following explanation is fairly
+informal. A more detailed look can be found in the appendix.
+
+Let's begin with a concrete example of an iterator type, like
+`std::vec::Iter<'a,T>`. We are interested in the projection of
+`Iterator::Item`:
+
+    <Iter<'a,T> as Iterator>::Item
+
+or, in the more succint (but potentially ambiguous) form:
+
+    Iter<'a,T>::Item
+
+Since I'm going to be talking a lot about this type, let's just call
+it `<PROJ>` for now. We would like to determine whether `<PROJ>: 'x` holds.
+
+Now, the easy way to solve `<PROJ>: 'x` would be to normalize `<PROJ>`
+by looking at the relevant impl:
+
+```rust
+impl<'b,U> Iterator for Iter<'b,U> {
+    type Item = &'b U;
+    ...
+}
+```
+
+From this impl, we can conclude that `<PROJ> == &'a T`, and thus
+reduce `<PROJ>: 'x` to `&'a T: 'x`, which in turn holds if `'a: 'x`
+and `T: 'x` (from the rule `OutlivesReference`).
+
+But often we are in a situation where we can't normalize the
```

Comment:
An example would be great. Is it simply any time you're working with a generic `I: Iterator`?